std/typecons.d

```diff
@@ -5893,3 +5893,143 @@ public:
     auto value = cast(int)flags_A;
     assert(value == Enum.A);
 }
+
+/**
+   Type constructor for head-const variables.
+
+   Head-const variables cannot be directly mutated or rebound, but references
+   reached through the variable are typed with their original mutability.
+
+   It is equivalent to $(D final) variables in D1 and Java,
+   as well as $(D readonly) variables in C#.
+ */
+struct HeadConst(T)
+    if(!is(T == const) && !is(T == immutable))
+{
+    private:
+    T data;
+
+    public:
+    /// $(D HeadConst) subtypes $(D T) as an rvalue.
+    inout(T) HeadConst_get() inout pure nothrow @nogc @safe
```

Comment:
> Is this actually a public function we want users to be able to use? I realize it probably needs to be public for the bellow alias to work, but let's not document it that way.

Yeah, it's public for AliasThis to work. I chose to document it because it's currently not possible to document an AliasThis declaration with DDoc, but it could simply be mentioned in the type description instead.

> This is not pure nothrow (etc...), as it depends on the attributes applied to T's postblit.

You're right, I was hoping the postblit operator would be called in the environment of the callsite :( Removing the explicit attributes and relying on inference seems to work well.

> Speaking of which, it might be worth having a const-by-ref overload.

I tried this, but the const-by-ref overload is never chosen when the this-reference is `const`, and when I make it `inout` like the other overload, I get "matches both" errors.